java/src/main/native/src/row_conversion.cu

```diff
@@ -187,6 +189,62 @@ struct batch_data {
   std::vector<row_batch> row_batches; // information about each batch such as byte count
 };
 
+rmm::device_uvector<size_type> build_string_row_offsets(table_view const &tbl,
+                                                        size_type fixed_width_and_validity_size,
+                                                        rmm::cuda_stream_view stream) {
+  auto const num_rows = tbl.num_rows();
+  rmm::device_uvector<size_type> d_row_offsets(num_rows, stream);
+
+  std::vector<strings_column_view::offset_iterator> string_columns_offsets;
+  auto column_ptrs = thrust::make_transform_iterator(
+      tbl.begin(), [](auto const &col) -> strings_column_view::offset_iterator {
+        if (!is_fixed_width(col.type())) {
+          CUDF_EXPECTS(col.type().id() == type_id::STRING, "only string columns are supported!");
+          strings_column_view scv = col;
+          return scv.offsets_begin();
+        } else {
+          return nullptr;
+        }
+      });
+  std::copy_if(column_ptrs, column_ptrs + tbl.num_columns(),
+               std::back_inserter(string_columns_offsets),
+               [](auto const &i) { return i != nullptr; });
+
+  auto const num_columns = static_cast<size_type>(string_columns_offsets.size());
+
+  auto d_string_columns_offsets = make_device_uvector_async(string_columns_offsets, stream);
+  auto key_iter = cudf::detail::make_counting_transform_iterator(
+      0, [num_columns] __device__(auto element_idx) { return element_idx / num_columns; });
+  auto data_iter = cudf::detail::make_counting_transform_iterator(
+      0, [d_string_columns_offsets = d_string_columns_offsets.data(), num_columns,
+          num_rows] __device__(auto element_idx) {
+        auto const row = element_idx / num_columns;
+        auto const col = element_idx % num_columns;
+
+        return d_string_columns_offsets[col][row + 1] - d_string_columns_offsets[col][row];
+      });
```

Comment:
Humm, from my perspective this computation is inefficient. You are looping col-by-col, then row by row. That means, for each row, you iteratively access all the cols before going to the next row.

Reply:
How about this:
```
auto const row = element_idx % num_rows;
auto const col = element_idx / num_rows;
...
```

Reply:
I'm not sure if this solution is more efficient. It should if we have large number of columns. Otherwise it may be slower.

Reply:
If you can have a benchmark to compare the solutions then it's great :smile:

Reply:
Spark max columns defaults to 100 and it seems far more likely to have a very large number of rows. With the requirement of keys being consecutive we can't simply flip the math. I will do some performance testing and report back.

Reply:
To be clear Spark does not have a maximum number of columns. It has a maximum number of columns before it stops trying to do code generation (because of limits in how the JVM compiler works). We follow that because we do code generation when going to rows. I don't think we follow that when going from rows to columns.

Generally we see a relatively small number of columns, especially compared to the number of rows being accessed. Also generally we don't see that many queries with massive numbers of columns in them.  If we do start to see more of that we are likely going to have to re-tune a lot of processing, not just here, to make that more efficient.

Reply:
I performance tested this code and it seems this function runs in about 1.2ms on my PC for 50 columns and 1,000,000 rows of intermixed ints and string columns. With the changes to not use `reduce_by_key` and march the data in a more natural way this time drops to 0.75ms. This seems worth it even though it removes the chance of the cool transform output iterator suggested in review. Thanks for pushing for this. I dismissed it probably because I was excited to use `reduce_by_key`.

Reply:
@revans2 is that default limit 100 or have I been led astray by my reading?

Reply:
Yes it is 100. 

https://github.com/apache/spark/blob/899d3bb44d7c72dc0179545189ac8170bde993a8/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L1356-L1362

But it actually calculates it with nesting levels too, so it is a bit more complicated.

https://github.com/apache/spark/blob/899d3bb44d7c72dc0179545189ac8170bde993a8/sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala#L566-L576

Reply:
> march the data in a more natural way this time drops to 0.75ms.

Kudos, @ttnghia and @hyperbolic2346!
I'm having a hard time grokking why this iteration order is faster. All the string columns have to eventually be accessed `num_rows` times. So this should be a matter of... proximity? All threads in the warp acting on proximal locations in memory?

Reply:
In the old way: we access rows `0` of all columns `0`, `1`, `2`, etc then we access rows `1` of all columns `0`, `1`, `2`, etc and so on.  Each row access will pull data from different columns from different locations in memory.
In the new way: we access rows `0`, `1`, `2`, etc of column `0`, then rows `0`, `1`, `2`, etc of column `1` and so on. So the data is pulled from contiguous memory locations.